File: rust/90.subsets-ii.rs
/*
 * @lc app=leetcode id=90 lang=rust
 *
 * [90] Subsets II
 */

// @lc code=start
impl Solution {
    pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut subsets = Vec::new();
        let mut nums = nums;
        nums.sort_unstable();
        fn backtrack(
            nums: &Vec<i32>,
            i: usize,
            subset: &mut Vec<i32>,
            subsets: &mut Vec<Vec<i32>>,
        ) {
            subsets.push(subset.clone());
            for j in i..nums.len() {
                if j > i && nums[j] == nums[j - 1] {
                    continue;
                }
                subset.push(nums[j]);
                backtrack(nums, j + 1, subset, subsets);
                subset.pop();
            }
        }
        backtrack(&nums, 0, &mut vec![], &mut subsets);
        subsets
    }
}
// @lc code=end

File: javascript/90.subsets-ii.js
/*
 * @lc app=leetcode id=90 lang=javascript
 *
 * [90] Subsets II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function (nums) {
  const subsets = [];
  nums.sort((a, b) => a - b);
  function backtrack(i, set) {
    subsets.push([...set]);
    for (let j = i; j < nums.length; j++) {
      if (j > i && nums[j] === nums[j - 1]) continue;
      set.push(nums[j]);
      backtrack(j + 1, set);
      set.pop();
    }
  }
  backtrack(0, []);
  return subsets;
};
// @lc code=end
